core/embed/bootloader/messages.c

```diff
@@ -649,13 +649,20 @@ int process_msg_FirmwareUpload(uint8_t iface_num, uint32_t msg_size,
 #endif
 
       uint32_t response = INPUT_CANCEL;
-      if (sectrue == is_new || sectrue == is_ilu) {
+      if (((vhdr.vtrust & VTRUST_ALL) == VTRUST_ALL) &&
+          (sectrue == is_new || sectrue == is_ilu)) {
         // new installation or interaction less updated - auto confirm
+        // only allowed for full-trust images
         response = INPUT_CONFIRM;
       } else {
-        int version_cmp = version_compare(hdr.version, current_hdr->version);
```

Comment:
How about this simplification:
```c
      if (((vhdr.vtrust & VTRUST_ALL) == VTRUST_ALL) &&
          (sectrue == is_new || sectrue == is_ilu)) {
        // new installation or interaction less updated - auto confirm
        // only allowed for full-trust images
        response = INPUT_CONFIRM;
      } else {
          int version_cmp = version_compare(hdr.version, current_hdr->version);
          response = ui_screen_install_confirm(
              &vhdr, &hdr, should_keep_seed || is_new, is_newvendor, is_new, version_cmp);
      }
```

Reply:
Rebased.

First, in the original version there was a problem of using `true` instead of `sectrue`.

Thus, the proposed code does `||` on `secbool` which is wrong. Also, the `current_hdr` is null when is_new is set so there is null pointer deref. 

With this taken into account, here is updated version https://github.com/trezor/trezor-firmware/pull/3709/commits/1148b43f91e88f601a9f7b1cd3f8e2cd89674869